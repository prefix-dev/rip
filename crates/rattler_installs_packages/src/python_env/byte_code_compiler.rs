use parking_lot::Mutex;
use std::collections::HashMap;
use std::sync::Arc;
use std::{
    io::{self, BufRead, BufReader, Write},
    path::{Path, PathBuf},
    process::{Command, Stdio},
    sync::mpsc::{channel, Sender},
};
use tempfile::NamedTempFile;
use thiserror::Error;

/// The result of compiling a Python script to byte code.
type CompilationResponse = Result<PathBuf, CompilationError>;
type CompilationRequest = PathBuf;

type BoxedCallback = Box<dyn FnOnce(CompilationResponse) + Send + 'static>;
type CompilationCallbackMap = HashMap<CompilationRequest, Vec<BoxedCallback>>;

/// An error that can occur when compiling a source file.
#[derive(Debug, Error, Clone)]
pub enum CompilationError {
    /// The file is not a python file.
    #[error("not a python file")]
    NotAPythonFile,

    /// The file could not be found.
    #[error("source file not found")]
    SourceNotFound,

    /// The file could not be compiled
    #[error("failed to compile")]
    FailedToCompile,

    /// The compilation host unexpectedly quit.
    #[error("host has quit")]
    HostQuit,
}

/// An error that can occur when spawning the compilation host.
#[derive(Debug, Error)]
pub enum SpawnCompilerError {
    /// Could not create the source code that runs in the compilation host.
    #[error("failed to create temporary file for compilation source")]
    FailedToCreateSource(#[source] io::Error),

    /// Failed to start the python executable
    #[error("failed to start python executable")]
    FailedToStartPython(#[source] io::Error),
}

/// An object that allows compiling python source code to byte code in a separate process.
pub struct ByteCodeCompiler {
    /// The channel that is used to send compilation requests to the compilation host. If this is
    /// dropped the attached thread will drop stdin of the child which will signal the child to
    /// exit.
    request_tx: Option<Sender<CompilationRequest>>,

    /// Callback functions per compilation request. These are called when the compilation host
    /// finishes processing a request.
    pending_callbacks: Arc<Mutex<Option<CompilationCallbackMap>>>,

    /// The child process. This is waited upon on drop.
    child: Option<std::process::Child>,

    // The file that contains the python source code of the compilation host. We keep this around
    // to make sure the file is not accidentally deleted while the compilation host is still using
    // it.
    _compilation_source: NamedTempFile,
}

impl ByteCodeCompiler {
    /// Constructs a new instance.
    ///
    /// This function spawns a new python process that will be used to compile python source code.
    pub fn new(python_path: &Path) -> Result<Self, SpawnCompilerError> {
        // Write the compilation host source code to a temporary file
        let compilation_source = tempfile::Builder::new()
            .prefix("pyc_compilation_host")
            .suffix(".py")
            .tempfile()
            .and_then(|mut f| {
                f.write_all(include_bytes!("compile_pyc.py"))?;
                Ok(f)
            })
            .map_err(SpawnCompilerError::FailedToCreateSource)?;

        // Start the compilation process
        let mut child = Command::new(python_path)
            .arg("-Wi")
            .arg("-u")
            .arg(compilation_source.path())
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .spawn()
            .map_err(SpawnCompilerError::FailedToStartPython)?;

        // Spawn a thread to process incoming compilation requests and forward them to the input of the
        // compilation process
        let (request_tx, request_rx) = channel::<CompilationRequest>();
        let mut child_stdin = child.stdin.take().expect("stdin is piped");
        std::thread::spawn(move || {
            while let Ok(request) = request_rx.recv() {
                tracing::trace!("compiling {}", request.display());
                if let Err(e) = child_stdin
                    .write_all(request.to_string_lossy().as_bytes())
                    .and_then(|_| child_stdin.write_all(b"\n"))
                    .and_then(|_| child_stdin.flush())
                {
                    tracing::error!("unexpected error writing to compilation host stdin: {}", e);
                    break;
                };
            }
        });

        // Spawn another thread to process the output of the compilation process and forward it to the
        // response channel.
        let pending_callbacks = Arc::new(Mutex::new(Some(CompilationCallbackMap::new())));
        let response_callbacks = pending_callbacks.clone();
        let child_stdout = BufReader::new(child.stdout.take().expect("stdout is piped"));
        std::thread::spawn(move || {
            #[derive(Debug, serde::Deserialize)]
            struct Response {
                path: PathBuf,
                output_path: Option<PathBuf>,
            }

            for line in child_stdout.lines() {
                match line.and_then(|line| Ok(serde_json::from_str::<Response>(&line)?)) {
                    Ok(response) => {
                        tracing::trace!("finished compiling '{}'", response.path.display());

                        let callbacks = {
                            let mut callback_lock = response_callbacks.lock();
                            let callbacks = callback_lock.as_mut().expect(
                                "the callbacks are not dropped until the end of this function",
                            );
                            callbacks.remove(&response.path)
                        };
                        match callbacks {
                            None => panic!(
                                "received a response for an unknown request '{}'",
                                response.path.display()
                            ),
                            Some(callbacks) => {
                                for callback in callbacks {
                                    callback(match &response.output_path {
                                        Some(output_path) => Ok(output_path.to_path_buf()),
                                        None => Err(CompilationError::FailedToCompile),
                                    })
                                }
                            }
                        };
                    }
                    Err(err) => {
                        tracing::error!(
                            "unexpected error reading from compilation host stdout: {}",
                            err
                        );
                        break;
                    }
                }
            }

            tracing::trace!("compilation host stdout closed");

            // Abort any pending callbacks and disable the ability to add new ones.
            let callbacks = response_callbacks
                .lock()
                .take()
                .expect("only we can drop the callbacks");
            for (_, callbacks) in callbacks {
                for callback in callbacks {
                    callback(Err(CompilationError::HostQuit))
                }
            }
        });

        Ok(Self {
            request_tx: Some(request_tx),
            pending_callbacks,
            child: Some(child),
            _compilation_source: compilation_source,
        })
    }

    /// Queue the compilation of the specified python file.
    ///
    /// The file is send to the compilation host which will immediately start compiling the file.
    /// The callback is called when the compilation is finished.
    ///
    /// Use the `drain` method to wait for the compilation to finish.
    pub fn compile<F: FnOnce(CompilationResponse) + Send + 'static>(
        &self,
        source_path: &Path,
        callback: F,
    ) -> Result<(), CompilationError> {
        if source_path.extension() != Some(std::ffi::OsStr::new("py")) {
            return Err(CompilationError::NotAPythonFile);
        }

        if !source_path.is_file() {
            return Err(CompilationError::SourceNotFound);
        }

        let mut lock = self.pending_callbacks.lock();
        let Some(callbacks) = lock.as_mut() else {
            return Err(CompilationError::HostQuit);
        };

        callbacks
            .entry(source_path.to_path_buf())
            .or_default()
            .push(Box::new(callback));

        self.request_tx
            .as_ref()
            .expect("the channel is only dropped on drop")
            .send(source_path.to_owned())
            .map_err(|_| CompilationError::HostQuit)
    }

    /// Compile the specified python file and wait for the compilation to finish.
    pub fn compile_and_wait(&self, source_path: &Path) -> Result<PathBuf, CompilationError> {
        let (tx, rx) = channel::<CompilationResponse>();
        self.compile(source_path, move |result| {
            tx.send(result)
                .expect("the reader is waiting on this response");
        })?;
        rx.recv()
            .expect("the sender was dropped before sending a response")
    }

    /// Wait for all queued compilations to finish.
    pub fn wait(mut self) -> Result<(), std::io::Error> {
        // Drop the request channel to signal the compilation host that we are done. This will
        // ensure that the stdin pipe of the compilation host is closed which will signal the host
        // to exit.
        drop(self.request_tx.take());

        // Wait for the compilation host to exit
        self.child
            .take()
            .expect("the child is only dropped on drop")
            .wait()?;

        Ok(())
    }
}

impl Drop for ByteCodeCompiler {
    fn drop(&mut self) {
        drop(self.request_tx.take());
        if let Some(mut child) = self.child.take() {
            child.wait().unwrap();
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::python_env::system_python_executable;

    #[test]
    fn test() {
        let python_path = system_python_executable().unwrap();

        // Create a temporary file that holds the sourcecode of the compilation host
        let mut compiler_source = tempfile::Builder::new().suffix(".py").tempfile().unwrap();
        compiler_source
            .write_all(include_bytes!("compile_pyc.py"))
            .unwrap();

        // Spawn a compiler and compile the compilation host source code.
        let compiler = ByteCodeCompiler::new(python_path).unwrap();
        let pyc_file = compiler.compile_and_wait(compiler_source.path()).unwrap();

        // Make sure the compiled file exists
        assert!(pyc_file.is_file(), "The compiled file does not exist!");

        // Wait should immediately return
        compiler.wait().unwrap();
    }

    #[test]
    fn test_failed_case() {
        let python_path = system_python_executable().unwrap();

        // Create a temporary file that holds the sourcecode of the compilation host
        let mut compiler_source = tempfile::Builder::new().suffix(".py").tempfile().unwrap();
        compiler_source.write_all(b"$").unwrap();

        // Spawn a compiler and compile the compilation host source code.
        let compiler = ByteCodeCompiler::new(python_path).unwrap();
        compiler
            .compile_and_wait(compiler_source.path())
            .unwrap_err();
    }
}
