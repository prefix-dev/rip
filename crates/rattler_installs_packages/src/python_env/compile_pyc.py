import sys
import json
import importlib
import compileall
from multiprocessing import Pool


def compile_one(path):
    success = compileall.compile_file(path, quiet=2, force=True)
    output_path = importlib.util.cache_from_source(path) if success else None
    return path, output_path


def compilation_finished(compilation_result):
    path, output_path = compilation_result
    print(json.dumps({"path": path, "output_path": output_path}))


if __name__ == "__main__":
    with sys.stdin:
        with Pool() as pool:
            while True:
                path = sys.stdin.readline().strip()
                if not path:
                    break
                pool.apply_async(compile_one, (path,), callback=compilation_finished)
