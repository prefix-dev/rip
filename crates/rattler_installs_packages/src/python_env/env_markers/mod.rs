use serde::{Deserialize, Serialize};
use std::ops::Deref;

mod from_env;

/// Describes the environment markers that can be used in dependency specifications to enable or
/// disable certain dependencies based on runtime environment.
///
/// Exactly the markers defined in this struct must be present during version resolution. Unknown
/// variables should raise an error.
///
/// Note that the "extra" variable is not defined in this struct because it depends on the wheel
/// that is being inspected.
///
/// The behavior and the names of the markers are described in PEP 508.
#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
#[allow(missing_docs)]
#[serde(transparent)]
pub struct Pep508EnvMakers(pub pep508_rs::MarkerEnvironment);

impl From<pep508_rs::MarkerEnvironment> for Pep508EnvMakers {
    fn from(value: pep508_rs::MarkerEnvironment) -> Self {
        Self(value)
    }
}

impl Deref for Pep508EnvMakers {
    type Target = pep508_rs::MarkerEnvironment;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
