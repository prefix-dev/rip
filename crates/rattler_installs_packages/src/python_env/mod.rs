//! Module for working with python environments.
//! Contains functionality for querying and manipulating python environments.

mod tags;

mod distribution_finder;

mod env_markers;

mod system_python;

mod uninstall;
mod venv;

pub use tags::{WheelTag, WheelTags};

pub use distribution_finder::{find_distributions_in_venv, Distribution, FindDistributionError};
pub use env_markers::Pep508EnvMakers;
pub(crate) use system_python::{
    system_python_executable, FindPythonError, ParsePythonInterpreterVersionError,
    PythonInterpreterVersion,
};
pub use uninstall::{uninstall_distribution, UninstallDistributionError};
pub(crate) use venv::{PythonLocation, VEnv};
