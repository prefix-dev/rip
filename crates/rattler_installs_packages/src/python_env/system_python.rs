use itertools::Itertools;
use std::io::ErrorKind;
use std::path::{Path, PathBuf};
use std::str::FromStr;
use thiserror::Error;

/// Error that can occur while finding the python executable.
#[derive(Debug, Error)]
pub enum FindPythonError {
    #[error("could not find python executable")]
    NotFound,
    #[error(transparent)]
    IoError(#[from] std::io::Error),
}

/// Try to find the python executable in the current environment.
/// Using sys.executable aproach will return original interpretator path
/// and not the shim in case of using which
pub fn system_python_executable() -> Result<PathBuf, FindPythonError> {
    // When installed with homebrew on macOS, the python3 executable is called `python3` instead
    // Also on some ubuntu installs this is the case
    // For windows it should just be python

    let output = match std::process::Command::new("python3")
        .arg("-c")
        .arg("import sys; print(sys.executable, end='')")
        .output()
        .or_else(|_| {
            std::process::Command::new("python")
                .arg("-c")
                .arg("import sys; print(sys.executable, end='')")
                .output()
        }) {
        Err(e) if e.kind() == ErrorKind::NotFound => return Err(FindPythonError::NotFound),
        Err(e) => return Err(FindPythonError::IoError(e)),
        Ok(output) => output,
    };

    let stdout = String::from_utf8_lossy(&output.stdout);
    let python_path = PathBuf::from_str(&stdout).unwrap();

    // sys.executable can return empty string or python's None
    if !python_path.exists() {
        return Err(FindPythonError::NotFound);
    }

    Ok(python_path)
}

/// Errors that can occur while trying to parse the python version
#[derive(Debug, Error)]
pub enum ParsePythonInterpreterVersionError {
    /// The version string is invalid.
    #[error("failed to parse version string, found '{0}' expect something like 'Python x.x.x'")]
    InvalidVersion(String),

    /// The Python interpreter could not be found when attempting to determine its version.
    #[error(transparent)]
    FindPythonError(#[from] FindPythonError),
}

/// Represents a Python interpreters version parts.
#[derive(Clone)]
pub struct PythonInterpreterVersion {
    /// The major version of the interpreter.
    pub major: u32,

    /// The minor version of the interpreter.
    pub minor: u32,

    /// The patch version of the interpreter.
    /// Also known as the "micro" version from Python's `sys.version_info`.
    pub patch: u32,
}

impl From<(u32, u32, u32)> for PythonInterpreterVersion {
    fn from(value: (u32, u32, u32)) -> Self {
        Self {
            major: value.0,
            minor: value.1,
            patch: value.2,
        }
    }
}

impl PythonInterpreterVersion {
    /// Get the version of the python interpreter
    /// Expects the string from `python --version` as input
    /// getting something along the lines of `Python 3.8.5`
    pub fn from_python_output(
        version_str: &str,
    ) -> Result<Self, ParsePythonInterpreterVersionError> {
        use ParsePythonInterpreterVersionError::InvalidVersion;

        // Split "Python 3.9.1" into "Python" and "3.9.1"
        let version_str = match version_str.split_once(' ') {
            Some(("Python", version)) => version,
            _ => return Err(InvalidVersion(version_str.to_owned())),
        };

        // Split the version into strings separated by '.' and parse them
        let parts = version_str
            .split('.')
            .map(str::trim)
            .map(FromStr::from_str)
            .collect::<Result<Vec<_>, _>>()
            .map_err(|_| InvalidVersion(version_str.to_owned()))?;

        // Extract the major, minor and patch version
        let Some((major, minor, patch)) = parts.into_iter().collect_tuple() else {
            return Err(InvalidVersion(version_str.to_owned()));
        };

        Ok(Self::new(major, minor, patch))
    }

    /// Creates a PythonInterpreterVersion from its constituent parts.
    pub fn new(major: u32, minor: u32, patch: u32) -> Self {
        Self {
            major,
            minor,
            patch,
        }
    }

    /// Get the python version from the system interpreter
    pub fn from_system() -> Result<Self, ParsePythonInterpreterVersionError> {
        Self::from_path(&system_python_executable()?)
    }

    /// Get the python version a path to the python executable
    pub fn from_path(path: &Path) -> Result<Self, ParsePythonInterpreterVersionError> {
        let output = std::process::Command::new(path)
            .arg("--version")
            .output()
            .map_err(|_| FindPythonError::NotFound)?;
        let version_str = String::from_utf8_lossy(&output.stdout);
        Self::from_python_output(&version_str)
    }
}

#[cfg(test)]
mod tests {
    use crate::python_env::PythonInterpreterVersion;

    #[test]
    pub fn parse_python_version() {
        let version = PythonInterpreterVersion::from_python_output("Python 3.8.5\n").unwrap();
        assert_eq!(version.major, 3);
        assert_eq!(version.minor, 8);
        assert_eq!(version.patch, 5);
    }
}
