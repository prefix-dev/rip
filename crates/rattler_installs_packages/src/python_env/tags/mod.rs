//! Wheels encode the Python interpreter, ABI, and platform that they support in their filenames
//! using platform compatibility tags. This module provides support for discovering what tags the
//! running Python interpreter supports and determining if a wheel is compatible with a set of tags.

mod from_env;

use indexmap::IndexSet;
use itertools::Itertools;
use serde_with::{DeserializeFromStr, SerializeDisplay};
use std::fmt::{Debug, Display, Formatter};
use std::str::FromStr;

/// A representation of a tag triple for a wheel.
#[derive(Debug, Clone, Hash, Eq, PartialEq, SerializeDisplay, DeserializeFromStr)]
pub struct WheelTag {
    /// The interpreter name, e.g. "py"
    pub interpreter: String,

    /// The ABI that a wheel supports, e.g. "cp37m"
    pub abi: String,

    /// The OS/platform the wheel supports, e.g. "win_am64".
    pub platform: String,
}

impl WheelTag {
    /// Parses a compound string into a `WheelTag`. A compound string is a string that contains
    /// multiple tags in a single string.
    ///
    /// ```rust
    /// # use rattler_installs_packages::python_env::WheelTag;
    /// let tags = WheelTag::from_compound_string(
    ///     "cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64").unwrap();
    ///
    /// assert_eq!(tags.len(), 2);
    /// assert_eq!(tags[0].interpreter, "cp310");
    /// assert_eq!(tags[0].abi, "cp310");
    /// assert_eq!(tags[0].platform, "manylinux_2_17_x86_64");
    /// assert_eq!(tags[1].interpreter, "cp310");
    /// assert_eq!(tags[1].abi, "cp310");
    /// assert_eq!(tags[1].platform, "manylinux2014_x86_64");
    ///
    /// ```
    pub fn from_compound_string(s: &str) -> Result<Vec<Self>, String> {
        let Some((interpreter, abi, platform)) =
            s.split('-').map(ToOwned::to_owned).collect_tuple()
        else {
            return Err(String::from("not enough '-' separators"));
        };

        Ok(interpreter
            .split('.')
            .cartesian_product(abi.split('.'))
            .cartesian_product(platform.split('.'))
            .map(|((interpreter, abi), platform)| Self {
                interpreter: interpreter.to_string(),
                abi: abi.to_string(),
                platform: platform.to_string(),
            })
            .collect())
    }
}

impl FromStr for WheelTag {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let Some((interpreter, abi, platform)) =
            s.split('-').map(ToOwned::to_owned).collect_tuple()
        else {
            return Err(String::from("not enough '-' separators"));
        };
        Ok(Self {
            interpreter,
            abi,
            platform,
        })
    }
}

impl Display for WheelTag {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}-{}-{}", &self.interpreter, &self.abi, &self.platform)
    }
}

/// Contains an ordered set of platform tags with which compatibility of wheels can be determined.
#[derive(Debug, Clone)]
pub struct WheelTags {
    tags: IndexSet<WheelTag>,
}

impl WheelTags {
    /// Returns an iterator over the supported tags.
    pub fn tags(&self) -> impl Iterator<Item = &'_ WheelTag> + '_ {
        self.tags.iter()
    }

    /// Determines the compatibility of the specified tag with the tags in this instance. Returns
    /// `None` if the specified tag is not compatible with any of the tags in this instance. Returns
    /// `Some(i)` where `i` indicates the compatibility level. The higher the number the more
    /// specific the tag is to the platform. The wheel artifact with the highest number should be
    /// preferred over others.
    pub fn compatibility(&self, tag: &WheelTag) -> Option<i32> {
        self.tags.get_index_of(tag).map(|score| -(score as i32))
    }

    /// Returns if the specified tag is compatible with this set.
    pub fn is_compatible(&self, tag: &WheelTag) -> bool {
        self.tags.contains(tag)
    }
}

impl FromIterator<WheelTag> for WheelTags {
    fn from_iter<T: IntoIterator<Item = WheelTag>>(iter: T) -> Self {
        Self {
            tags: FromIterator::from_iter(iter),
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_from_str() {
        let tag = WheelTag::from_str("py2-none-any").unwrap();
        assert_eq!(tag.interpreter, "py2");
        assert_eq!(tag.abi, "none");
        assert_eq!(tag.platform, "any");
    }
}
