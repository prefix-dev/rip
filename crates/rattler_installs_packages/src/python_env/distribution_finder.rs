//! This module implements logic to locate so called python distributions (installed artifacts)
//! in an environment.
//!
//! The implementation is based on the <https://packaging.python.org/en/latest/specifications/recording-installed-packages>
//! which is based on [PEP 376](https://peps.python.org/pep-0376/) and [PEP 627](https://peps.python.org/pep-0627/).

use crate::artifacts::wheel::InstallPaths;
use crate::python_env::WheelTag;
use crate::{types::NormalizedPackageName, types::PackageName, types::RFC822ish};
use indexmap::IndexSet;
use itertools::Itertools;
use pep440_rs::Version;
use serde::{Deserialize, Serialize};
use std::{
    ffi::OsStr,
    path::{Path, PathBuf},
    str::FromStr,
};
use thiserror::Error;

/// Information about a distribution found by `find_distributions_in_venv`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Distribution {
    /// The name of the distribution
    pub name: NormalizedPackageName,

    /// The version of the distribution
    pub version: Version,

    /// The installer that was responsible for installing the distribution
    pub installer: Option<String>,

    /// The path to the .dist-info directory relative to the root of the environment.
    pub dist_info: PathBuf,

    /// The specific tags of the distribution that was installed or `None` if this information
    /// could not be retrieved.
    pub tags: Option<IndexSet<WheelTag>>,
}

/// An error that can occur when running `find_distributions_in_venv`.
#[derive(Debug, Error)]
pub enum FindDistributionError {
    /// An IO error occurred
    #[error(transparent)]
    IoError(#[from] std::io::Error),

    /// Failed to parse a WHEEL file
    #[error("failed to parse '{0}'")]
    FailedToParseWheel(PathBuf, #[source] <RFC822ish as FromStr>::Err),

    /// Failed to parse WHEEL tags
    #[error("failed to parse wheel tag {0}")]
    FailedToParseWheelTag(String),
}

/// Locates the python distributions (packages) that have been installed in the virtualenv rooted at
/// `root`.
pub fn find_distributions_in_venv(
    root: &Path,
    paths: &InstallPaths,
) -> Result<Vec<Distribution>, FindDistributionError> {
    // We will look for distributions in the purelib/platlib directories
    let locations = [paths.purelib(), paths.platlib()]
        .into_iter()
        .map(|p| root.join(p))
        .unique()
        .filter(|p| p.is_dir())
        .collect_vec();

    // Iterate over all the entries in the in the locations and look for .dist-info entries.
    let mut result = Vec::new();
    for location in locations {
        for entry in location.read_dir()? {
            let entry = entry?;
            if entry.file_type()?.is_dir() {
                if let Some(dist) = analyze_distribution(entry.path())? {
                    result.push(Distribution {
                        dist_info: pathdiff::diff_paths(&dist.dist_info, root)
                            .unwrap_or(dist.dist_info),
                        ..dist
                    })
                }
            }
        }
    }

    Ok(result)
}

/// Analyzes a `.dist-info` directory to see if it actually contains a python distribution (package).
fn analyze_distribution(
    dist_info_path: PathBuf,
) -> Result<Option<Distribution>, FindDistributionError> {
    let Some((name, version)) = dist_info_path
        .file_name()
        .and_then(OsStr::to_str)
        .and_then(|n| n.strip_suffix(".dist-info"))
        .and_then(|n| n.split_once('-'))
    else {
        // If we are unable to parse the distribution name we just skip.
        return Ok(None);
    };

    // Check if the METADATA file is present. This is the only file that is mandatory, so if its
    // missing this folder is not a python distribution.
    if !dist_info_path.join("METADATA").is_file() {
        return Ok(None);
    }

    // Parse the name
    let Ok(name) = PackageName::from_str(name) else {
        // If the package name cannot be parsed, just skip
        return Ok(None);
    };

    // Parse the version
    let Ok(version) = Version::from_str(version) else {
        // If the version cannot be parsed, just skip
        return Ok(None);
    };

    // Try to read the INSTALLER file from the distribution directory
    let installer = std::fs::read_to_string(dist_info_path.join("INSTALLER"))
        .map(|i| i.trim().to_owned())
        .ok();

    // Check if there is a WHEEL file from where we can read tags
    let wheel_path = dist_info_path.join("WHEEL");
    let tags = if wheel_path.is_file() {
        let mut parsed = RFC822ish::from_str(&std::fs::read_to_string(&wheel_path)?)
            .map_err(move |e| FindDistributionError::FailedToParseWheel(wheel_path, e))?;

        Some(
            parsed
                .take_all("Tag")
                .into_iter()
                .map(|tag| {
                    WheelTag::from_compound_string(&tag)
                        .map_err(|_| FindDistributionError::FailedToParseWheelTag(tag))
                })
                .flatten_ok()
                .collect::<Result<IndexSet<_>, _>>()?,
        )
    } else {
        None
    };

    Ok(Some(Distribution {
        dist_info: dist_info_path,
        name: name.into(),
        version,
        installer,
        tags,
    }))
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_find_distributions() {
        // Describe the virtual environment
        let venv_path =
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../test-data/find_distributions/");
        let install_paths = InstallPaths::for_venv((3, 8, 5), true);

        // Find all distributions
        let mut distributions = find_distributions_in_venv(&venv_path, &install_paths).unwrap();

        // Sort to get consistent ordering across platforms
        distributions.sort_by(|a, b| a.name.cmp(&b.name));

        insta::assert_ron_snapshot!(distributions, {
            "[].dist_info" => insta::dynamic_redaction(move |value, _path| {
                value.as_str().unwrap().replace('\\', "/")
            }),
        });
    }
}
