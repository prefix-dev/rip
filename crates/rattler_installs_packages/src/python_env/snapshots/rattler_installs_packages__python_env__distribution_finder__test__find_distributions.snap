---
source: crates/rattler_installs_packages/src/python_env/distribution_finder.rs
expression: distributions
---
[
  Distribution(
    name: "click",
    version: "7.1.2",
    installer: Some("pip"),
    dist_info: "Lib/site-packages/click-7.1.2.dist-info",
    tags: Some([
      "py2-none-any",
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "flask",
    version: "1.1.4",
    installer: Some("pip"),
    dist_info: "Lib/site-packages/Flask-1.1.4.dist-info",
    tags: Some([
      "py2-none-any",
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "itsdangerous",
    version: "1.1.0",
    installer: Some("pip"),
    dist_info: "Lib/site-packages/itsdangerous-1.1.0.dist-info",
    tags: Some([
      "py2-none-any",
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "jinja2",
    version: "2.11.3",
    installer: Some("pip"),
    dist_info: "Lib/site-packages/Jinja2-2.11.3.dist-info",
    tags: Some([
      "py2-none-any",
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "markupsafe",
    version: "1.1.1",
    installer: Some("pip"),
    dist_info: "Lib/site-packages/MarkupSafe-1.1.1.dist-info",
    tags: Some([
      "cp35-cp35m-win_amd64",
    ]),
  ),
  Distribution(
    name: "pip",
    version: "9.0.1",
    installer: Some("pip"),
    dist_info: "Lib/site-packages/pip-9.0.1.dist-info",
    tags: Some([
      "py2-none-any",
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "setuptools",
    version: "28.8.0",
    installer: Some("pip"),
    dist_info: "Lib/site-packages/setuptools-28.8.0.dist-info",
    tags: Some([
      "py2-none-any",
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "werkzeug",
    version: "1.0.1",
    installer: Some("pip"),
    dist_info: "Lib/site-packages/Werkzeug-1.0.1.dist-info",
    tags: Some([
      "py2-none-any",
      "py3-none-any",
    ]),
  ),
]
